instrumentation/kafka-clients/kafka-clients-0.11/library/src/main/java/io/opentelemetry/instrumentation/kafka/TracingKafkaUtils.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.kafka;
+
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.opentelemetry.javaagent.instrumentation.kafka.KafkaHeadersRefGetter;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.function.BiFunction;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.header.Headers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("rawtypes")
+public class TracingKafkaUtils {
+
+  private static final Logger logger = LoggerFactory.getLogger(TracingKafkaUtils.class);
+  public static final String TO_PREFIX = "To_";
+  public static final String FROM_PREFIX = "From_";
+
+  private static final TextMapGetter<Headers> HEADERS_TEXT_MAP_GETTER =
+      new KafkaHeadersRefGetter();
+
+  private static final TextMapSetter<Headers> HEADERS_TEXT_MAP_SETTER =
+      (carrier, key, value) -> carrier.add(key, value.getBytes(StandardCharsets.UTF_8));
+
+  static Tracer getTracer() {
+    return GlobalOpenTelemetry.getTracer(StandardSpanDecorator.COMPONENT_NAME);
+  }
+
+  private static TextMapPropagator propagator() {
+    return GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+  }
+
+  /**
+   * Extract Context from record headers.
+   *
+   * @param headers record headers
+   * @return span context
+   */
+  public static Context extractSpanContext(Headers headers) {
+    return propagator().extract(Context.current(), headers, HEADERS_TEXT_MAP_GETTER);
+  }
+
+  /**
+   * Inject current Context to record headers.
+   *
+   * @param headers record headers
+   */
+  public static void inject(Headers headers) {
+    inject(Context.current(), headers);
+  }
+
+  /**
+   * Inject Context to record headers.
+   *
+   * @param context the context
+   * @param headers record headers
+   */
+  public static void inject(Context context, Headers headers) {
+    propagator().inject(context, headers, HEADERS_TEXT_MAP_SETTER);
+  }
+
+  public static <K, V> Span buildAndInjectSpan(ProducerRecord<K, V> record) {
```

Comment:
@mateuszrzeszutek OK, will try to rewrite stuff to use `Instrumenter` API

Can I also move some of the extractors, setters, etc to `kafka-clients-common`?
Since that looks like a better place ... if not to even create a whole new library module under `kafka-clients-common`?

Reply:
Sure, please do! I put some commonly used consumer-related code there because I needed it before, but it's definitely *the* place to put common producer-related code (or just commons kafka instrumentation code).

Also, that module probably needs a little correction: it's `kafka-clients-common:javaagent` right now, but since you're using it in a library instrumentation it should probably be renamed to `kafka-clients-common:library` (the `javaagent` part should be removed from package names too).

Anyway, thanks for tackling that - if you run into any issues/questions with the Instrumenter API please don't hesitate and ask - either here or in our slack channel.